README.md
# 6502 CPU emulator

A 6502 CPU emulator created during a Dojo.

+ http://www.8bitpickles.com/writing-a-cpu-emulator-as-a-dojo/
+ [6502 Datasheeet](http://archive.6502.org/datasheets/wdc_w65c02s_oct_8_2018.pdf)

## Example

```clojure
(def program
  "LDA 100 ; foo
   ADC 7
   STA 15
   BRK"
  )

(def mem (load-program (asm program)))

(def cpu {:pc  0
          :ar  0
          :brk false})

(dump!! (run {:cpu cpu :mem mem}))

pc  = 0x07    7
ar  = 0x6b  107
brk = true
00000000  a9 64 69 07 8d 0f 00 00  00 00 00 00 00 00 00 6b  |.di............k|
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000f0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
=> nil
```



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-04-09
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-04-09
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/cpu/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/cpu/compare/0.1.0...0.1.1


test/cpu/core_test.clj
(ns cpu.core-test
  (:require [clojure.test :refer :all]
            [cpu.core :refer :all]))

(deftest program-tests
  (testing "Simple program adds 2 numbers and stores in memory"
    (is (=
          {:cpu {:pc 7, :ar 107, :xr 0, :yr 0, :brk true, :eq false}
           :mem [169 100
                 105 7
                 141 15
                 0 0
                 0 0
                 0 0
                 0 0
                 0 107]}
          (run
            {:cpu {:pc 0, :ar 0, :xr 0, :yr 0, :brk false, :eq false}
             :mem [169 100
                   105 7
                   141 15
                   0 0
                   0 0
                   0 0
                   0 0
                   0 0]})))))


src/cpu/debug.clj
(ns cpu.debug)

(def mem-row-format
  (str "%08x  "
       "%02x %02x %02x %02x %02x %02x %02x %02x  "
       "%02x %02x %02x %02x %02x %02x %02x %02x  "
       "|%c%c%c%c%c%c%c%c%c%c%c%c%c%c%c%c|"))

(defn is-ascii?
  [chr]
  (and (>= chr 32)
       (<= chr 126)))

(defn long->ascii
  [chr]
  (byte
    (if (is-ascii? chr)
      chr
      46)))

(defn format-mem-row
  [row]
  (let [loc   (first row)
        data  (second row)
        chars (map long->ascii data)]
    (apply (partial format mem-row-format)
           (flatten (list loc data chars)))))

(defn group-memory-into-rows
  ([mem]
   (group-memory-into-rows mem 16))
  ([mem width]
   (map-indexed #(list (reduce + %2) (* width %) %2)
                (partition width mem))))

(defn dedup-mem
  [mem]
  (let [rows (group-memory-into-rows mem)
        end  (second (last rows))]
    (reverse
      (reduce (fn [s e]
                (let [row-sum      (first e)
                      last-row-sum (first (first s))
                      pos          (second e)]
                  (if (and (= last-row-sum row-sum)
                           (not (= pos end)))
                    s
                    (conj s e))))
              '() rows))))

(defn format-mem
  [mem]
  (let [lines (dedup-mem mem)]
    (map #(format-mem-row (rest %)) lines)))

(defn format-register
  [[k v]]
  (format "%-4s = 0x%02x % 4d" k v v))

(defn format-flags
  [[k v]]
  (format "%-4s = %b" k v))

(defn format-registers
  [cpu]
  (map format-register (select-keys cpu [:pc :ar :xr :yr :sp])))

(defn format-status-flag
  [cpu]
  (map format-flags (select-keys cpu [:brk :eq])))

(defn format-dump
  [vm]
  (flatten
    (list (format-registers (:cpu vm))
          (format-status-flag (:cpu vm))
          (format-mem (:mem vm)))))

;

(defn dump!!
  [vm]
  (doall (map println (format-dump vm)))
  nil)

(defn dump!
  [vm]
  (dump!! vm)
  vm)


src/cpu/helpers.clj
(ns cpu.helpers)

(def max-mem 640)

(defn inc-pc
  ([pc]
   (mod (inc pc) max-mem))
  ([pc n]
   (mod (+ n pc) max-mem)))

(defn cpu-pc
  [vm]
  (get-in vm [:cpu :pc]))

(defn set-pc
  [vm value]
  (assoc-in vm [:cpu :pc] value))

(defn cpu-brk
  [vm]
  (get-in vm [:cpu :brk]))

(defn cpu-ar
  [vm]
  (get-in vm [:cpu :ar]))

(defn cpu-xr
  [vm]
  (get-in vm [:cpu :xr]))

(defn cpu-yr
  [vm]
  (get-in vm [:cpu :yr]))

(defn cpu-brk-flag?
  [vm]
  (= true (cpu-brk vm)))

(defn read-mem
  [vm address]
  (nth (:mem vm) address))

(defn write-mem
  [vm address value]
  (assoc-in vm [:mem address] value))

(defn next-opcode
  [vm]
  (read-mem vm (cpu-pc vm)))

(defn incr-pc
  ([vm]
   (let [pc (inc-pc (cpu-pc vm))]
     (assoc-in vm [:cpu :pc] pc)))
  ([vm n]
   (let [pc (inc-pc (cpu-pc vm) n)]
     (assoc-in vm [:cpu :pc] pc))))

(defn set-ar
  [vm n]
  (assoc-in vm [:cpu :ar] n))

(defn inc-ar
  [vm n]
  (let [ar (cpu-ar vm)]
    (set-ar vm (+ ar n))))

(defn set-xr
  [vm n]
  (assoc-in vm [:cpu :xr] n))

(defn incr-xr
  [vm]
  (let [xr (cpu-xr vm)]
    (set-xr vm (inc xr))))

(defn set-yr
  [vm n]
  (assoc-in vm [:cpu :yr] n))

(defn incr-yr
  [vm]
  (let [yr (cpu-yr vm)]
    (set-yr vm (inc yr))))


(defn set-flag-brk
  [vm]
  (assoc-in vm [:cpu :brk] true))

(defn set-flag-eq
  [vm equal?]
  (assoc-in vm [:cpu :eq] equal?))

(defn cpu-flag-eq
  [vm]
  (get-in vm [:cpu :eq]))

(defn cpu-sp
  [vm]
  (get-in vm [:cpu :sp]))

(defn set-sp
  [vm value]
  (assoc-in vm [:cpu :sp] value))

(defn inc-sp
  [vm]
  (let [sp (cpu-sp vm)]
    (set-sp vm (+ 1 sp))))

(defn dec-sp
  [vm]
  (let [sp (cpu-sp vm)]
    (set-sp vm (dec sp))))

(defn make-mem
  [mem size]
  (let [padding (- size (.length mem))]
    (vec (into mem (repeat padding 0)))))

(defn load-program
  ([program]
   (make-mem program max-mem))
  ([program mem-size]
   (make-mem program mem-size)))

(defn lookup-instruction
  [opcode instructions]
  (first (filter #(= opcode (:opcode %)) instructions)))



src/cpu/assemble.clj
(ns cpu.assemble
  (:require [clojure.string :as str]
            [clojure.edn :as edn]
            [cpu.instructions :refer :all]))

(defn split-lines
  [^CharSequence lines]
  (map str/triml (str/split-lines lines)))

(defn ^String strip-comments
  [^CharSequence line]
  (str/trimr
    (if-let [idx (str/index-of line ";")]
      (. line (subSequence 0 idx))
      line)))

(defn parse-value
  [value]
  (edn/read-string value))

(defn make-instruction
  [opcode value]
  {:opcode (keyword (.toLowerCase opcode))
   :value (parse-value value)})

(defn tokenize-line
  [line]
  (map str/trim (str/split line #" ")))

(defn parse-line
  [^CharSequence line]
  (let [tokens (tokenize-line line)
        opcode (first tokens)
        operand (second tokens)]
    (make-instruction opcode operand)))

(defn opcode->byte
  [opcode]
  (let [instruction (first (filter #(= opcode (:name %)) @instructions))]
    (if (nil? instruction)
      0x00
      (:opcode instruction))))

(defn assemble-code
  [code]
  (for [instruction code]
    (let [opcode-byte (opcode->byte (:opcode instruction))
          value (:value instruction)]
      (if (nil? value)
        opcode-byte
        [opcode-byte value]))))

(defn parse-program
  [^CharSequence program]
  (for [line (split-lines program)]
    (-> line
        (strip-comments)
        (parse-line))))

(defn asm
  [^CharSequence program]
  (-> program
      (parse-program)
      (assemble-code)
      (flatten)
      (vec)))



src/cpu/instructions.clj
(ns cpu.instructions
  (:require [cpu.helpers :refer :all]))

(def instructions (atom []))

(defn- make-instruction
  "Create a new instruction"
  [name doc [opcode size] func]
  (let []
    {:name   name
     :doc    doc
     :opcode opcode
     :size   size
     :func   func}))

(defn defasm
  "Define a new instruction"
  [name doc [opcode size] func]
  (let [inst (make-instruction name doc [opcode size] func)]
    (swap! instructions conj inst)))

(defasm :brk
        "Stop the program"
        [0x00 1]
        (fn [vm]
          (-> vm
              (set-flag-brk)
              (incr-pc))))

(defasm :lda
        "Load Accumulator with memory"
        [0xa9 2]
        (fn [vm]
          (let [pc    (cpu-pc vm)
                value (read-mem vm (+ 1 pc))]
            (-> vm
                (set-ar value)
                (incr-pc 2)))))

(defasm :adc
        "Add memory to accumulator with Carry"
        [0x69 2]
        (fn [vm]
          (let [pc    (cpu-pc vm)
                value (read-mem vm (+ 1 pc))]
            (-> vm
                (inc-ar value)
                (incr-pc 2)))))

(defasm :sta
        "Store the accumulator in memory"
        [0x8d 2]
        (fn [vm]
          (let [pc      (cpu-pc vm)
                ar      (cpu-ar vm)
                address (read-mem vm (+ 1 pc))]
            (-> vm
                (write-mem address ar)
                (incr-pc 2)))))

(defasm :ldx
        "Load the X register with memory"
        [0xa2 2]
        (fn [vm]
          (let [pc    (cpu-pc vm)
                value (read-mem vm (+ 1 pc))]
            (-> vm
                (set-xr value)
                (incr-pc 2)))))

(defasm :inx
        "Increment X register by one"
        [0xe8 1]
        (fn [vm]
          (-> vm
              (incr-xr)
              (incr-pc))))

(defasm :iny
        "Increment Y register by one"
        [0xc8 1]
        (fn [vm]
          (-> vm
              (incr-yr)
              (incr-pc))))

(defasm :cpy
        "Compare memory and Y register"
        [0xe0 2]
        (fn [vm]
          (let [pc    (cpu-pc vm)
                yr    (cpu-yr vm)
                value (read-mem vm (+ 1 pc))]
            (-> vm
                (set-flag-eq (= yr value))
                (incr-pc 2)))))

(defasm :bne
        "Branch if not equal Pz=0"
        [0xd0 2]
        (fn [vm]
          (let [pc     (cpu-pc vm)
                value  (read-mem vm (+ 1 pc))
                pc-inc (if (not (cpu-flag-eq vm))
                         value
                         2)]
            (-> vm
                (incr-pc pc-inc)))))

(defasm :stax
        "Store the value from the A register in the memory location stored in the X register"
        ;; This is a non-standard instruction
        ;; I picked 0x8f because I think it's unused
        [0x8f 1]
        (fn [vm]
          (let [ar (cpu-ar vm)
                xr (cpu-xr vm)]
            (-> vm
                (write-mem xr ar)
                (incr-pc)))))

(defasm :dey
        "Decrement Y register by one"
        [0x88 1]
        (fn [vm]
          (let [yr (cpu-yr vm)]
            (-> vm
                (set-yr (dec yr))
                (incr-pc)))))

(defasm :ldy
        "Load the Y register with memory"
        [0xa0 2]
        (fn [vm]
          (let [pc    (cpu-pc vm)
                value (read-mem vm (+ 1 pc))]
            (-> vm
                (set-yr value)
                (incr-pc 2)))))

(defasm :nop
        "No operation"
        [0xea 1]
        (fn [vm]
          (-> vm
              (incr-pc))))

(defasm :jsr
        "Jump to new location Saving Return"
        [0x20 2]
        (fn [vm]
          (let [address    (cpu-sp vm)
                value      (cpu-pc vm)
                next-value (read-mem vm (inc (cpu-pc vm)))]
            (-> vm
                (write-mem address value)
                (dec-sp)
                (set-pc (dec next-value))))))

(defasm :ret
        "Return from subroutine"
        [0x60 1]
        (fn [vm]
          (let [sp             (cpu-sp vm)
                return-address (read-mem vm (inc sp))]
            (-> vm
                (inc-sp)
                (set-pc (inc return-address))
                (incr-pc)))))

src/cpu/core.clj
(ns cpu.core
  (:gen-class)
  (:require [cpu.debug :refer :all]
            [cpu.helpers :refer :all]
            [cpu.instructions :refer :all]
            [cpu.assemble :refer :all]))

(defn step
  ([vm]
   (step vm false))
  ([vm debug?]
   (let [opcode      (next-opcode vm)
         instruction (lookup-instruction opcode @instructions)
         func        (if (not (nil? instruction))
                       (:func instruction)
                       (fn [vm] vm))]
     (if debug?
       (func (dump! vm))
       (func vm)))))

(defn run
  ([vm]
   (run vm false))
  ([vm debug?]
   (loop [vm (step vm debug?)]
     (if (not (cpu-brk-flag? vm))
       (recur (step vm debug?))
       vm))))

;
; Main
;

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Hello, World!"))


;
; Test
;

(def program-1
  "LDA 100 ; foo
   ADC 7
   STA 15
   LDX 12
   INX
   LDY 24
   BRK"
  )

(def program-2
  "LDX 128      ; 0
  LDA 0x77      ; 2
  STAX          ; 4
  INX           ; 5
  LDA 0x68      ; 6
  STAX          ; 8
  INX           ; 9
  LDA 0x6F      ; 10
  STAX          ; 12
  INX           ; 13
  LDA 0x20      ; 14
  STAX          ; 16
  INX           ; 17
  LDA 0x6c      ; 18
  STAX          ; 20
  INX           ; 21
  LDA 0x65      ; 22
  STAX          ; 24
  INX           ; 25
  LDA 0x74      ; 26
  STAX          ; 28
  INX           ; 29
  LDA 0x20      ; 30
  STAX          ; 32
  INX           ; 33
  LDA 0x74      ; 34
  STAX          ; 36
  INX           ; 37
  LDA 0x68      ; 38
  STAX          ; 40
  INX           ; 41
  LDA 0x65      ; 42
  STAX          ; 44
  INX           ; 45
  LDA 0x20      ; 46
  STAX          ; 48
  INX           ; 49
  LDA 0x64      ; 50
  STAX          ; 52
  INX           ; 53
  LDA 0x6F      ; 54
  STAX          ; 56
  INX           ; 57
  LDA 0x67      ; 58
  STAX          ; 60
  INX           ; 61
  LDA 0x73      ; 62
  STAX          ; 64
  INX           ; 65
  LDA 0x20      ; 66
  STAX          ; 68
  INX           ; 69
  LDA 0x6F      ; 70
  STAX          ; 72
  INX           ; 73
  LDA 0x75      ; 74
  STAX          ; 76
  INX           ; 77
  LDA 0x74      ; 78
  STAX          ; 80
  INX           ; 81
  LDA 0x20      ; 82
  STAX          ; 84
  INX           ; 85
  LDY 3         ; 86
  LDA 0x77      ; 88
  STAX          ; 90
  INX           ; 91
  LDA 0x68      ; 92
  STAX          ; 94
  INX           ; 95
  LDA 0x6F      ; 96
  STAX          ; 98
  INX           ; 99
  LDA 0x20      ; 100
  STAX          ; 102
  INX           ; 103
  DEY           ; 104
  CPY 0         ; 105
  BNE -19       ; 107
  BRK           ; 109")

(def program-3
  "LDX 128
  LDA 0x77
  STAX
  INX
  LDA 0x68
  STAX
  INX
  LDA 0x6F
  STAX
  INX
  LDA 0x20
  STAX
  INX
  LDA 0x6c
  STAX
  INX
  LDA 0x65
  STAX
  INX
  LDA 0x74
  STAX
  INX
  LDA 0x20
  STAX
  INX
  LDA 0x74
  STAX
  INX
  LDA 0x68
  STAX
  INX
  LDA 0x65
  STAX
  INX
  LDA 0x20
  STAX
  INX
  LDA 0x64
  STAX
  INX
  LDA 0x6F
  STAX
  INX
  LDA 0x67
  STAX
  INX
  LDA 0x73
  STAX
  INX
  LDA 0x20
  STAX
  INX
  LDA 0x6F
  STAX
  INX
  LDA 0x75
  STAX
  INX
  LDA 0x74
  STAX
  INX
  LDA 0x20
  STAX
  INX
  LDY 3     ; loop (y = 3)
  JSR 97    ;   call 'write who' subroutine
  DEY       ;   y--
  CPY 0     ;   test y==0
  BNE -5    ; while (y != 0)
  BRK
  LDA 0x77  ; start 'write who' subroutine
  STAX
  INX
  LDA 0x68
  STAX
  INX
  LDA 0x6F
  STAX
  INX
  LDA 0x20
  STAX
  INX
  RET       ; end 'write who' subroutine")

(def mem (load-program (asm program-1)))

(def cpu {:pc  0
          :ar  0
          :xr  0
          :yr  0
          :sp  (dec max-mem)
          :brk false
          :eq  false})

(def vm {:cpu cpu :mem mem})

;(dump!! (run vm))


